apps/balancer-tools/src/app/apr/(utils)/tokenYield.ts

```diff
@@ -20,42 +20,124 @@ export const publicClient = createPublicClient({
 });
 
 const SECONDS_IN_DAY = 86400;
+const DAYS_IN_YEAR = 365;
+const SECONDS_IN_YEAR = DAYS_IN_YEAR * SECONDS_IN_DAY;
 
 // eslint-disable-next-line @typescript-eslint/no-unused-vars
-export async function getTokenAprByPoolId(poolId: string) {
-  const { currentRate, rate24hAgo } = await getDailyRateDiff();
+export async function getPoolTokensApr(
+  chain: string,
+  poolId: `0x${string}`
+) {
+  const rateProviders = await getPoolTokensRateProviders(chain, poolId);
 
-  return [
+  const now = Math.round(new Date().getTime() / 1000);
+
+  return await Promise.all(rateProviders
+    .filter(({ address }) => address !== zeroAddress)
+    .map(
+      async ({
+        address: rateProviderAddress,
+        token: { symbol, address: tokenAddress },
+      }) => ({
+        address: tokenAddress,
+        symbol,
+        yield: await getAPRFromRateProviderInterval(
+          rateProviderAddress,
+          now - SECONDS_IN_DAY,
+          now
+        ),
+      })
+    ))
+}
+
+async function getAPRFromRateProviderInterval(
+  rateProviderAddress: `0x${string}`,
+  timeStart: number,
+  timeEnd: number
+) {
+  const { endRate, startRate } = await getIntervalRates(
+    rateProviderAddress,
+    timeStart,
+    timeEnd
+  );
+
+  return getAPRFromRate(startRate, endRate, timeStart, timeEnd);
+}
+
+
+function getAPRFromRate(
+  rateStart: number,
+  rateEnd: number,
+  timeStart: number,
+  timeEnd: number
+) {
+  const duration = timeEnd - timeStart;
+  const rateOfReturn = (rateEnd - rateStart) / rateStart;
+  const annualScalingFactor = SECONDS_IN_YEAR / duration;
+
+  const APR = rateOfReturn * annualScalingFactor * 100;
+
+  return APR;
+}
+
+async function getPoolTokensRateProviders(
+  chain: string,
+  poolId: `0x${string}`
+) {
+  const data = await pools.gql(String(chain)).PoolRateProviders({ poolId });
+  if (!data.pool?.priceRateProviders?.length) return [
     {
-      address: "0xae78736cd615f374d3085123a210448e74fc6393",
-      symbol: "rETH",
-      yield: ((currentRate - rate24hAgo) / 10 ** 18) * 365 * 100,
-    },
+      address: "0x1a8F81c256aee9C640e14bB0453ce247ea0DFE6F",
+      token: {
+        address: "0xae78736cd615f374d3085123a210448e74fc6393",
+        symbol: "rETH",
+      },
+    }
   ];
+
+  return data.pool?.priceRateProviders;
 }
```

Comment:
I ended up getting to [this](https://bleu-llc.slack.com/archives/C05FE3PLPP0/p1694639017259739) issue with the subgraph and decided not to pursue it further. If the subgraph value is not reliable though we can always fallback to fetching this directly onchain